src/core/src/SO3Utils.cpp

```diff
@@ -0,0 +1,173 @@
+        return false;
+    }
+
+    if (weights.size() && (inputRotations.size() != weights.size()))
+    {
+        iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", "Vectors size mismatch: weights and inputRotations must be same size.");
+        return false;
+    }
+
+    if (options.tolerance < 0.0)
+    {
+        iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", "The tolerance is supposed to be a positive number.");
+        return false;
+    }
+
+    if (options.stepSize < 0.0)
+    {
+        iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", "The stepSize is supposed to be non-negative.");
+        return false;
+    }
+
+    double nrRot = inputRotations.size();
+
+    double totalWeight = 0.0;
+    if (weights.size())
+    {
+        for (size_t i = 0; i < weights.size(); ++i)
+        {
+            if (weights[i] < 0.0)
+            {
+                std::stringstream ss;
+                ss << "The weight at index " << i << " is negative.";
+                iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", ss.str().c_str());
+                return false;
+            }
+            totalWeight += weights[i];
+        }
+    }
+    else
+    {
+        totalWeight = nrRot;
+    }
+
+    if (totalWeight < options.tolerance)
+    {
+        iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", "The sum of all the weights is lower than the tolerance.");
+        return false;
+    }
+
+    // initial condition for optimization
+    meanRotation = inputRotations[0];
+
+    bool optimal_R_found{false};
+    auto start = std::chrono::high_resolution_clock::now();
+    int iteration = 1;
+    while (!optimal_R_found)
+    {
+        // Riemannian gradient descent with constant step size
+        iDynTree::AngularMotionVector3 perturbation;
+        perturbation.zero();
+        Eigen::Map<Eigen::Vector3d> r(toEigen(perturbation));
+
+        for (size_t idx = 0; idx < nrRot; ++idx)
+        {
+            const iDynTree::Rotation& R_i = inputRotations[idx];
+            double w_i = weights.size() ? weights[idx] : 1.0;
+
+            r =  r + w_i * toEigen((meanRotation.inverse() * R_i).log());
+        }
+        r = r * (1.0 / totalWeight);
+
+        double norm = r.norm();
+
+        if (std::isinf(norm))
+        {
+            iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", "Inf detected.");
+            return false;
+        }
+
+        if (std::isnan(norm))
+        {
+            iDynTree::reportError("SO3Utils", "geodesicL2WeightedMeanRotation", "Nan detected.");
+            return false;
+        }
+
+        if (r.norm() <= options.tolerance)
+        {
+            optimal_R_found = true;
+            continue;
+        }
+
+
+        r *= options.stepSize;
+
+        meanRotation = meanRotation * perturbation.exp();
+        auto now = std::chrono::high_resolution_clock::now();
```

Comment:
There is nothing in the standard that ensures that `std::chrono::high_resolution_clock` is monotonic (see https://github.com/WebAssembly/binaryen/issues/766 and https://en.cppreference.com/w/cpp/chrono/high_resolution_clock). This means that if this clock for some reason  is reset, the timeout logic will break. I suggest to use `steady_clock` for this kind of uses: https://en.cppreference.com/w/cpp/chrono/steady_clock .

Reply:
Done with https://github.com/robotology/idyntree/pull/651/commits/fd9f17ca3da741e869513ce616d42c9b40ff9360